lib/memjs/memjs.js

```diff
@@ -170,6 +186,137 @@ Client.prototype.get = function(key, callback) {
   });
 };
 
+/** Build a pipelined get multi request by sending one GETKQ for each key (quiet, meaning it won't respond if the value is missing) followed by a no-op to force a response (and to give us a sentinel response that the pipeline is done)
+ *
+ * cf https://github.com/couchbase/memcached/blob/master/docs/BinaryProtocol.md#0x0d-getkq-get-with-key-quietly
+ */
+Client.prototype._buildGetMultiRequest = function(keys) {
+  // start at 24 for the no-op command at the end
+  var requestSize = 24;
+  for (var keyIdx in keys) {
+    requestSize += Buffer.byteLength(keys[keyIdx], 'utf8') + 24;
+  }
+
+  var request = Buffer.alloc(requestSize);
+  request.fill();
+
+  this.incrSeq();
+
+  var bytesWritten = 0;
+  for (keyIdx in keys) {
+    var key = keys[keyIdx];
+    bytesWritten += addToRequestBuffer(constants.OP_GETKQ, key, '', '', this.seq, request, bytesWritten);
+  }
+
+  bytesWritten += addToRequestBuffer(
+    constants.OP_NO_OP, '', '', '', this.seq, request, bytesWritten);
+
+  return request;
+};
+
+/** Executing a pipelined (multi) get against a single server. This is a private implementation detail of getMulti. */
+Client.prototype._getMultiToServer = function (serv, keys, callback) {
+  var self = this;
+
+  var responseMap = {};
+
+  var handle = function(response) {
+    switch (response.header.status) {
+    case  0:
+      if (callback) {
+        var deserialized = self.serializer.deserialize(response.header.opcode, response.val, response.extras);
+        // When we get the no-op response, we are done with this one getMulti in the per-backend fan-out
+        if (response.header.opcode === constants.OP_NO_OP) {
+          // this ensures the handler will be deleted from the responseCallbacks map in server.js
+          handle.quiet = true;
+          callback(null, responseMap);
+        } else {
+          var key = response.key.toString();
+          responseMap[key] = deserialized.value;
+        }
+      }
+      break;
+    case 1:
+      if (callback) { callback(null, null, null); }
+      break;
+    default:
+      var errorMessage = 'MemJS GET: ' + errors[response.header.status];
+      self.options.logger.log(errorMessage);
+      if (callback) { callback(new Error(errorMessage), null, null); }
+    }
+  };
+  // This prevents the handler from being deleted
+  // after the first response. Logic in server.js.
+  handle.quiet = true;
+
+  var request = this._buildGetMultiRequest(keys);
+  serv.onResponse(this.seq, handle);
+  this.incrSeq();
+  serv.onError(this.seq, function(err) {
+    if (callback) { callback(err, serv.host + ':' + serv.port, null); }
```

Comment:
Likewise - do we need to do something here with `handle.quiet = false` so we de-alloc the handle func?

Reply:
(Update: no - onError is only ever called from a codepath that calls all the error handlers, removes all listeners, and destroys the socket)